contact_map/concurrence.py

```diff
@@ -0,0 +1,246 @@
+import itertools
+import mdtraj as md
+import numpy as np
+
+try:
+    import matplotlib.pyplot as plt
+except ImportError:
+    HAS_MATPLOTLIB = False
+else:
+    HAS_MATPLOTLIB = True
+
+
+class Concurrence(object):
+    """Superclass for contact concurrence objects.
+
+    Contact concurrences measure what contacts occur simultaneously in a
+    trajectory. When defining states, one usually wants to characterize
+    based on multiple contacts that are made simultaneously; contact
+    concurrences makes it easier to identify those.
+
+    Parameters
+    ----------
+        values : list of list of bool
+            the whether a contact is present for each contact pair at each
+            point in time; outer list is length number of frames, inner list
+            in length number of (included) contacts
+        labels : list of string
+            labels for each contact pair
+    """
+    def __init__(self, values, labels=None):
+        self.values = values
+        self.labels = labels
+
+    # @property
+    # def lifetimes(self):
+        # pass
+
+    def set_labels(self, labels):
+        """Set the contact labels
+
+        Parameters
+        ----------
+        labels : list of string
+            labels for each contact pair
+        """
+        self.labels = labels
+
+    def __getitem__(self, label):
+        idx = self.labels.index(label)
+        return self.values[idx]
+
+    # temporarily removed until we find a good metric here; this metric did
+    # not seem optimal and I stopped using it, so remove from code before
+    # release (can add back in later)
+    # def coincidence(self, label_list):
+        # this_list = np.asarray(self[label_list[0]])
+        # coincidence_list = this_list
+        # norm_sq = sum(this_list)
+        # for label in label_list[1:]:
+            # this_list = np.asarray(self[label])
+            # coincidence_list &= this_list
+            # norm_sq *= sum(this_list)
+
+        # return sum(coincidence_list) / np.sqrt(norm_sq)
+
+
+class AtomContactConcurrence(Concurrence):
+    """Contact concurrences for atom contacts.
+
+    Parameters
+    ----------
+    trajectory : :class:`mdtraj.Trajectory`
+        the trajectory to analyze
+    atom_contacts : list
+        output from ``contact_map.atom_contacts.most_common()``
+    cutoff : float
+        cutoff, in nm. Should be the same as used in the contact map.
+    """
+    def __init__(self, trajectory, atom_contacts, cutoff=0.45):
+        # TODO: the use of atom_contacts as input from most_common is weird
+        atom_pairs = [[contact[0][0].index, contact[0][1].index]
+                      for contact in atom_contacts]
+        labels = [str(contact[0]) for contact in atom_contacts]
+        distances = md.compute_distances(trajectory, atom_pairs=atom_pairs)
+        vector_f = np.vectorize(lambda d: d < cutoff)
+        values = list(map(list, zip(*vector_f(distances))))
```

Comment:
please add a comment for this magic

Reply:
and this could be changed into: `list(map(list, vector_f(distances).T)` 
as you do want to map the transpose of the array to a list and I suspect this is more performant as it used numpy instead of zip. But I didn't check my suspicion